### PR TITLE
Add support for graphql-transport-ws protocol

### PR DESCRIPTION
**Pull request by maximilianmaihoefner**
Quick and dirty implementation of the graqhql-ws-transport protocol, as mentioned in #958 

Fixes https://github.com/zino-hofmann/graphql-flutter/issues/958

Reference implementation https://github.com/enisdenjo/graphql-ws

### CODE REVIEW

File: packages/graphql/lib/src/links/websocket_link/websocket_link.dart
Comment (jamesgorman2) [inline packages/graphql/lib/src/links/websocket_link/websocket_link.dart:44]: For `graphql-transport-ws`, connect needs to block before subscribe is sent, otherwise a 4401 error will be returned.

```
  void connectOrReconnect() async {
    _socketClient?.dispose();
    _socketClient = await SocketClient(
      url,
      config: config,
      protocol: subProtocol,
    ).connect();
  }
```

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart
Comment (jamesgorman2) [inline packages/graphql/lib/src/links/websocket_link/websocket_client.dart:170]: Replace `_connect` call in constructor with public `connect` call by invoker to allow blocking.

Comment (jamesgorman2) [inline packages/graphql/lib/src/links/websocket_link/websocket_client.dart:170]: For graphql-transport-ws, connect needs to block before subscribe is sent, otherwise a 4401 error will be returned.

Replace `_connect` call in constructor with public `connect` call by invoker to allow blocking.

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart
Comment (jamesgorman2) [inline packages/graphql/lib/src/links/websocket_link/websocket_client.dart:251]: Need to wait for ack.

```
      if (protocol == SocketSubProtocol.graphqlTransportWs) {
        // wait for ack
        await _messages.firstWhere((message) => message.type != MessageTypes.connectionAck);
      }
```

Comment (jamesgorman2) [inline packages/graphql/lib/src/links/websocket_link/websocket_client.dart:251]: Need to wait for response to init. Eg,

```
      if (protocol == SocketSubProtocol.graphqlTransportWs) {
        // wait for ack
        await _messages.firstWhere((message) => message.type != MessageTypes.connectionAck);
      }
```

**Review by jamesgorman2 — CHANGES_REQUESTED** [review]
_Edited for clarity_

For `graphql-transport-ws`, `_connect` needs to block before subscribe is sent otherwise, a `4401 Unauthorized` error will be returned. 

Per [graphql-ws/PROTOCOL](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#connectioninit), the required flow is `ConnectionInit` followed by `ConnectionAck` before any gql messages are sent.

Suggestions for code changed in-line. 

If you follow the full suggestion, the new signature for `_connect` is 
```
Future<SocketClient> connect() async
```
and needs to return `this`.

Tested using latest flutter web and `apollo-server-express` + `graphql-ws` on Win 10.

**Comment by vincenzopalazzo** [comment]
Rebased on master, and I will review in this week with a deep drive to the spec https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md

Sorry for delay here!

File: packages/graphql/lib/src/links/websocket_link/websocket_link.dart
Comment (vincenzopalazzo) [reply packages/graphql/lib/src/links/websocket_link/websocket_link.dart:44]: Mh I take a long reflection on this site and I think you have right! I usually don't use the lock when I write a client for the client because multiple requests can have multiple responses.

However, when we insert this logic inside a flutter app, this will be a little bit messy!

I will try to move this code with the lock, but I will take a little bit of time to reason on it. I'm 100% sure of this choice!

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart
Comment (vincenzopalazzo) [reply packages/graphql/lib/src/links/websocket_link/websocket_client.dart:170]: I agree with this spec fix, but I'm not able why move the constructor to the public, can you add more reason for this choice?

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart
Comment (vincenzopalazzo) [inline packages/graphql/lib/src/links/websocket_link/websocket_client.dart:488]:
```suggestion
GraphQLSocketMessage operation =  StartOperation(
            id,
            serialize(payload),
          );
          if (protocol == SocketSubProtocol.graphqlTransportWs) {
            operation = SubscribeOperation(
                id,
                serialize(payload),
              );
          }
```

**Review by vincenzopalazzo — COMMENTED** [review]
LGTM

I have only some stylistic comments, but I know that this PR is old, and I can work on this if you have no time.

I will need another review with another spec reading, and in addition, I need to think about how test it!

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart
Comment (budde377) [inline packages/graphql/lib/src/links/websocket_link/websocket_client.dart:463]:
```suggestion
            .whereType<SubscriptionNext>()
```

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart
Comment (budde377) [inline packages/graphql/lib/src/links/websocket_link/websocket_client.dart:256]: All these conditionals are making the code harder to maintain and reason about. If we could refactor the protocol strategy out somehow, e.g. in a subclass, that would be stellar!

File: packages/graphql/lib/src/links/websocket_link/websocket_link.dart
Comment (budde377) [reply packages/graphql/lib/src/links/websocket_link/websocket_link.dart:44]: Rather than using a mutex, you can probably just store the future (or maybe `FutureOr`) in `_socketClient` instead of the `SocketClient`.

```dart
Future<SocketClient>? _socketClient;
```

then you'd await that future in `request`.

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart
Comment (vincenzopalazzo) [reply packages/graphql/lib/src/links/websocket_link/websocket_client.dart:256]: Yeah, this is was I was thinking! I'm thinking only about how to divide the strategy pattern to include also a state pattern that manages the connection, so in this case, we can avoid all the if-else inside the class.

If the author of the PR have not time, I will take this task

**Comment by jamesgorman2** [comment]
> Rather than using a mutex, you can probably just store the future (or maybe FutureOr) in _socketClient instead of the SocketClient.

I think this is the better solution in general, but there may still be a race between `connectOrReconnect` and `dispose` with the code as-is. Looking back, it should probably be pulled out into its own issue/PR after the core issue here is addressed (the reentrancy issue is a low likelihood/low impact leak). 

There is some simplification that can occur since
```dart
  // cannot be final because we're changing the instance upon a header change.
  SocketClient? _socketClient;
...
  /// Connects or reconnects to the server with the specified headers.
  Future<void> connectOrReconnect() async {
```
doesn't hold for the current implementation as the headers can't be changed. This means that a single future and some simple semaphores in link & client should work and be simpler.

**Comment by budde377** [comment]
@jamesgorman2 

> I think this is the better solution in general, but there may still be a race between connectOrReconnect and dispose with the code as-is

With the explicitly calling connect and assigning the future client synchronously, can you go into detail what kind of race conditions you’re seeing?

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart
Comment (maximilianmaihoefner) [reply packages/graphql/lib/src/links/websocket_link/websocket_client.dart:256]: @vincenzopalazzo Right now I'm pretty stomped at work and don't have much time. Feel free to pick up where I left off!

**Comment by vincenzopalazzo** [comment]
>I think this is the better solution in general, but there may still be a race between connectOrReconnect and dispose with the code as-is

I'm not able to imagine your race condition inside one client, I'm so sorry. The client should be able to send multiple requests at a time, and the server is responsible for the race condition. Of curse, putting a mutex everywhere could be good from the race condition point of view, however, for now without having a use case where the race condition is a real problem I considered the use of mutex a bad design.

Of curse, if you call a mutation and at the same time call `connectOrReconnect` it is a race condition, but it is a usage error!

P.S: I'm just not able to see the problem, but happy to reconsider it with a real use case or workflow that can happen inside an app.

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart
Comment (vincenzopalazzo) [reply packages/graphql/lib/src/links/websocket_link/websocket_client.dart:256]: @maximilianmaihoefner ok! thanks for your initial work and sorry if I start to review this so late!

**Comment by jamesgorman2** [comment]
The race is a low likelihood resource bleed: 
* only in the case of passing up future errors
* regardless of using a future socket client or not
* without restructuring the method flow

Screenshot 2022-05-03 184020

In this sequence, `SC1` is disposed twice and `SC2` becomes detached and cannot be disposed.

Two options are:
* don't wait for `_socketClient.connect()` or `_socketClient.dispose()` - you can't pass errors out but they are safe w.r.t. calling subscribe
* use futures and value swaps, something like this flows on all the errors and obeys the dispose comment:

```dart
/// A Universal Websocket [Link] implementation to support the websocket transport.
/// It supports subscriptions, query and mutation operations as well.
///
/// NOTE: the actual socket connection will only get established after a [Request] is handled by this [WebSocketLink].
/// If you'd like to connect to the socket server instantly, call the [connect] method after creating this [WebSocketLink] instance.
class WebSocketLink extends Link {
  /// Creates a new [WebSocketLink] instance with the specified config.
  WebSocketLink(
    this.url, {
    this.config = const SocketClientConfig(),
    this.subProtocol = SocketSubProtocol.graphqlTransportWs,
  });

  final String url;
  final String subProtocol;
  final SocketClientConfig config;

  bool disposed = false;

  // cannot be final because we're late binding.
  Future<SocketClient>? _socketClient;

  @override
  Stream<Response> request(Request request, [forward]) async* {
    if (disposed) {
      yield* Stream.error('Cannot request on disposed WebSocketLink');
      return;
    }

    final sc = await connect();
    yield* sc.subscribe(request, true);
  }

  /// Connects to the server with the specified headers.
  Future<SocketClient> connect() async {
    if (disposed) {
      return Future.error('Cannot connect to disposed WebSocketLink');
    }
    _socketClient ??= SocketClient(
        url,
        config: config,
        protocol: subProtocol,
      ).connect();
    return _socketClient!;
  }

  /// Disposes the underlying socket client explicitly. Only use this, if you want to disconnect from
  /// the current server in favour of another one. If that's the case, create a new [WebSocketLink] instance.
  @override
  Future<void> dispose() async {
    disposed = true;
    final fsc = _socketClient;
    _socketClient = null;
    final sc  = await fsc;
    return sc?.dispose() ?? Future.value();
  }
}
```

**Comment by budde377** [comment]
@jamesgorman2 thanks for explaining. In your example, you seem to be re-connecting on every request, is this intentional?

Alternatively, I think we can get by the race by making `connectOrReconnect` and `dispose` completely synchronous and then effectively queueing the operations. Something along the lines of:

```dart
  Future<SocketClient?> _socketClientF = Fututre.value();

  Future<SocketClient> connectOrReconnect() {
    return connectF = _socketClientF
      .then(c => c?.dispose())
      .then(() => SocketClient(url, config: config).connect());
    _socketClientF = connectF.catchError((_e) => null);
    return connectF;
  }

  Future<SocketClient> get _client => _socketClient.then((c) => c  ?? connectOrReconnect());

  /// Disposes the underlying socket client explicitly. Only use this, if you want to disconnect from
  /// the current server in favour of another one. If that's the case, create a new [WebSocketLink] instance.
  Future<void> dispose() {
    const disposeF = _socketClient
       .then((c) => c?.dispose());
    _socketClientF = disposeF.catchError((_e) => null);
    return disposeF;
  }

  @override
  Stream<Response> request(Request request, [forward]) async* {
    SocketClient client = await _client; // Possibly yield error here?
    yield* client.subscribe(request, true);
  }

```

**Comment by jamesgorman2** [comment]
> you seem to be re-connecting on every request, is this intentional?

`connect` is behaving as `connectIfNotConnected` via the `??=`, so will only connect to the underlying socket once. (This was a language recommendation from Idea - it replaced the original if null test I had). I'll edit it to make it clearer. 

I'm also removing the ambiguity around reconnection and disposal in the current implementation, which influences my design decisions. 

> Alternatively, I think we can get by the race by making connectOrReconnect and dispose completely synchronous

Yes, this also works, and encapsulates the async code. Also like that you've highlighteda different failed connection behaviour. 

Comparing these makes me think there are questions that should be clarified:
* can a link be meaningfully reconnected?
* should a link be able to reconnect after disposal?
* should a failed connection be allowed to retry?
* should connection errors propagate up?

**Comment by budde377** [comment]
> connect is behaving as connectIfNotConnected via the ??=, so will only connect to the underlying socket once. (This was a language recommendation from Idea - it replaced the original if null test I had). I'll edit it to make it clearer.

You're right! Thanks for clarifying.

> Comparing these makes me think there are questions that should be clarified:
> 
> * can a link be meaningfully reconnected?
> * should a link be able to reconnect after disposal?
> * should a failed connection be allowed to retry?
> * should connection errors propagate up?

I think these are great questions! It's not obvious to me why we should support re-connection after disposing. Instead, we should encourage users to just create another link. Arguably, I believe we should retry on requests and yield the errors as part of the response to the request. This follows the standard query and mutations where server and connection errors are returned as part of the query response. 

So ideally, we could remove the reconnect feature as proposed by your solution. The problem is however that it would introduce a breaking change to the API. Which ultimately might delay the release of this feature (i.e., for a major version we most likely want to bundle up a few breaking features). So arguably, we should try and solve the race with the existing API, even though it might not be the best solution.

**Comment by jamesgorman2** [comment]
@budde377 thank you, exactly the sort of answers I was looking for.

@vincenzopalazzo are you still able to look at this? 
 I'm about to get slammed with a bunch of things, that'll tie me up for a week or two, but I'll be able to review & answer questions.

**Comment by vincenzopalazzo** [comment]
>are you still able to look at this?

I put it on my to-do list and see what I can do, but I think the race condition in the context of a follow-up request, in this PR I'm interested in the was protocol support.

I will try to split this talk into one or more issues so we can keep track

**Comment by jamesgorman2** [comment]
> I put it on my to-do list and see what I can do, but I think the race condition in the context of a follow-up request, in this PR I'm interested in the was protocol support.
> 
> I will try to split this talk into one or more issues so we can keep track

No complaints from me about breaking it up.

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart
Comment (vincenzopalazzo) [reply packages/graphql/lib/src/links/websocket_link/websocket_client.dart]: Fixed in https://github.com/maximilianmaihoefner/graphql-flutter/tree/69e46832999656979bf2c6829efc972087c01dff

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart
Comment (vincenzopalazzo) [reply packages/graphql/lib/src/links/websocket_link/websocket_client.dart:170]: Fixed in https://github.com/maximilianmaihoefner/graphql-flutter/tree/69e46832999656979bf2c6829efc972087c01dff

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart
Comment (vincenzopalazzo) [reply packages/graphql/lib/src/links/websocket_link/websocket_client.dart]: Fixed in https://github.com/zino-hofmann/graphql-flutter/pull/1060/commits/7d4939fa67fb5c42b46749865909c8812d2fcba0

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart
Comment (vincenzopalazzo) [reply packages/graphql/lib/src/links/websocket_link/websocket_client.dart:269]: >All these conditionals are making the code harder to maintain and reason about. If we could refactor the protocol strategy out somehow, e.g. in a subclass, that would be stellar!

@budde377 do you think that this change request can be a follow PR, in this way we can put in beta the new protocol support? and in the meanwhile working in the refactoring

**Comment by vincenzopalazzo** [comment]
IMHO this PR is ready to land, I divided the other topics discussed inside the following issue:

- https://github.com/zino-hofmann/graphql-flutter/issues/1124
- https://github.com/zino-hofmann/graphql-flutter/issues/1123
- https://github.com/zino-hofmann/graphql-flutter/issues/1122


When @budde377 will review it, I'm happy to merge!

**Comment by budde377** [comment]
@vincenzopalazzo good idea on postponing the other issues on another PR.

What are your thoughts on testing this?

**Comment by budde377** [comment]
> @vincenzopalazzo good idea on postponing the other issues on another PR.
> 
> What are your thoughts on testing this?

@vincenzopalazzo, NVM, just saw #1124

**Review by budde377 — APPROVED** [review]

**Comment by vincenzopalazzo** [comment]
>NVM, just saw https://github.com/zino-hofmann/graphql-flutter/issues/1124

Yeah! just thinking what is the best way to do this integration testing, looks like we don't have a global mock server to test graphql client implementation. I'm curious how Appollo handle this integration testing

**Review by vincenzopalazzo — APPROVED** [review]

**Pull request merged**
